### ISSUE TITLE
NGINX Proxy Manager // bash: line 58: nvm: command not found

### ISSUE THREAD

**Issue by point-source** [tteck/Proxmox#1709]
### A clear and concise description of the issue.

When trying to use the nginx proxy manager script on Proxmox 8.0.3, I am getting this error related to nodejs failing to install:

```
+ msg_info 'Installing Node.js'
+ local 'msg=Installing Node.js'
+ echo -ne ' - \033[33mInstalling Node.js...'
 - Installing Node.js...+ bash /dev/fd/63
++ curl -fsSL https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.3/install.sh
=> Downloading nvm from git to '/root/.nvm'
=> Cloning into '/root/.nvm'...
remote: Enumerating objects: 360, done.
remote: Counting objects: 100% (360/360), done.
remote: Compressing objects: 100% (306/306), done.
remote: Total 360 (delta 40), reused 169 (delta 28), pack-reused 0
Receiving objects: 100% (360/360), 220.28 KiB | 2.62 MiB/s, done.
Resolving deltas: 100% (40/40), done.
* (HEAD detached at FETCH_HEAD)
  master
=> Compressing and cleaning up git repository

=> Appending nvm source string to /root/.profile
=> bash_completion source string already in /root/.profile
=> Close and reopen your terminal to start using nvm or run the following to use it now:

export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
+ . /root/.bashrc
+ nvm install 16.20.1
bash: line 58: nvm: command not found
++ error_handler 58 '$STD nvm install 16.20.1'
++ local exit_code=127
++ local line_number=58
++ local 'command=$STD nvm install 16.20.1'
++ local 'error_message=\033[01;31m[ERROR]\033[m in line \033[01;31m58\033[m: exit code \033[01;31m127\033[m: while executing command \033[33m$STD nvm install 16.20.1\033[m'
++ echo -e '\n\033[01;31m[ERROR]\033[m in line \033[01;31m58\033[m: exit code \033[01;31m127\033[m: while executing command \033[33m$STD nvm install 16.20.1\033[m'

[ERROR] in line 58: exit code 127: while executing command $STD nvm install 16.20.1
++ [[ 58 -eq 24 ]]
+ exit
```

I have searched the past issues and have tried the following settings to resolve:
- Using default settings
- Clearing browser cache
- Using a different browser
- Debian 11
- Debian 12
- Ubuntu 23.04
- Restarting the proxmox host

When I use ubuntu, I get the `/root/.bashrc: line 6: PS1: unbound variable` error. All other times, I get the error posted here.

### Please provide detailed steps to reproduce the issue.

1. Run the nginxproxymanager script with any settings on a proxmox 8 host

**Comment by tteck** [comment]
Are you running the command from an SSH terminal or directly within the Proxmox shell?

**Comment by point-source** [comment]
Primarily browser shell. I did test ssh once and got the warning but proceeded anyway. Here was the output of the ssh attempt:

```
+ msg_info 'Installing Node.js'
+ local 'msg=Installing Node.js'
+ echo -ne ' - \033[33mInstalling Node.js...'
 - Installing Node.js...+ bash /dev/fd/63
++ curl -fsSL https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.3/install.sh
curl: (28) SSL connection timeout
+++ error_handler 56 'curl -fsSL https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.3/install.sh'
+++ local exit_code=28
+++ local line_number=56
+++ local 'command=curl -fsSL https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.3/install.sh'
+++ local 'error_message=\033[01;31m[ERROR]\033[m in line \033[01;31m56\033[m: exit code \033[01;31m28\033[m: while executing command \033[33mcurl -fsSL https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.3/install.sh\033[m'
+++ echo -e '\n\033[01;31m[ERROR]\033[m in line \033[01;31m56\033[m: exit code \033[01;31m28\033[m: while executing command \033[33mcurl -fsSL https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.3/install.sh\033[m'
+++ [[ 56 -eq 24 ]]
/dev/fd/63: line 2: $'\E[01': command not found
/dev/fd/63: line 2: $'31m[ERROR]\E[m': command not found
/dev/fd/63: line 2: $'31m56\E[m:': command not found
/dev/fd/63: line 2: $'31m28\E[m:': command not found
++ error_handler 56 '$STD bash <(curl -fsSL https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.3/install.sh)'
++ local exit_code=127
++ local line_number=56
++ local 'command=$STD bash <(curl -fsSL https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.3/install.sh)'
++ local 'error_message=\033[01;31m[ERROR]\033[m in line \033[01;31m56\033[m: exit code \033[01;31m127\033[m: while executing command \033[33m$STD bash <(curl -fsSL https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.3/install.sh)\033[m'
++ echo -e '\n\033[01;31m[ERROR]\033[m in line \033[01;31m56\033[m: exit code \033[01;31m127\033[m: while executing command \033[33m$STD bash <(curl -fsSL https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.3/install.sh)\033[m'

[ERROR] in line 56: exit code 127: while executing command $STD bash <(curl -fsSL https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.3/install.sh)
++ [[ 56 -eq 24 ]]
+ exit
```

**Comment by tteck** [comment]
The script runs without any problems from this end. 

```
    _   __      _               ____                           __  ___                                 
   / | / /___ _(_)___  _  __   / __ \_________ __  ____  __   /  |/  /___ _____  ____ _____ ____  _____
  /  |/ / __  / / __ \| |/_/  / /_/ / ___/ __ \| |/_/ / / /  / /|_/ / __  / __ \/ __  / __  / _ \/ ___/
 / /|  / /_/ / / / / />  <   / ____/ /  / /_/ />  </ /_/ /  / /  / / /_/ / / / / /_/ / /_/ /  __/ /    
/_/ |_/\__, /_/_/ /_/_/|_|  /_/   /_/   \____/_/|_|\__, /  /_/  /_/\__,_/_/ /_/\__,_/\__, /\___/_/     
      /____/                                      /____/                            /____/             
 
Using Advanced Settings
Using Distribution: debian
Using debian Version: 12
Using Container Type: 1
Using Root Password: Automatic Login
Container ID: 112
Using Hostname: nginxproxymanager
Using Disk Size: 4
Allocated Cores: 1
Allocated RAM: 1024
Using Bridge: vmbr0
Using IP Address: dhcp
Using Gateway IP Address: Default
Disable IPv6: yes
Using Interface MTU Size: Default
Using DNS Search Domain: Host
Using DNS Server IP Address: Host
Using MAC Address: AE:1A:60:07:B7:59
Using Vlan: Default
Enable Root SSH Access: no
Enable Verbose Mode: no
Creating a Nginx Proxy Manager LXC using the above advanced settings
 ✓ Using local for Template Storage.
 ✓ Using local-lvm for Container Storage.
 ✓ Updated LXC Template List
 ✓ LXC Container 112 was successfully created.
 ✓ Started LXC Container
 ✓ Set up Container OS
 ✓ Network Connected: 192.168.86.253 
 ✓ Internet Connected
 ✓ DNS Resolved github.com to 140.82.114.4
 ✓ Updated Container OS
 ✓ Installed Dependencies
 ✓ Updated Python
 ✓ Installed Openresty
 ✓ Installed Node.js
 ✓ Installed Yarn
 ✓ Downloaded Nginx Proxy Manager v2.10.4
 ✓ Set up Enviroment
 ✓ Built Frontend
 ✓ Initialized Backend
 ✓ Created Service
 ✓ Customized Container
 ✓ Started Services
 ✓ Cleaned
 ✓ Completed Successfully!

Nginx Proxy Manager should be reachable by going to the following URL.
         http://192.168.86.253:81 
```

**Comment by point-source** [comment]
Hmm, which proxmox version?

**Comment by tteck** [comment]
8.0.5

**Comment by point-source** [comment]
Okay looks like I'm behind a bit (8.0.3). Also, would blocking third party cookies / other privacy measures impact this? What browser capabilities are used by the script? What assumptions are made about the environment?

EDIT: I don't see a version 8.0.5. Is this pre-release? My version appears to be latest stable as of today.

**Comment by point-source** [comment]
I just did a test and installed nvm manually while the script was running but before it got to the node part. This allowed the node.js installation to succeed. So the error is correct in that it is not finding nvm. But I can't figure it why.

**Comment by tteck** [comment]
I suspect that the problem could be related to reloading the `.bashrc` file (line 57: `. ~/.bashrc`).

**Comment by tteck** [comment]
As this issue is unrelated to the script, it is considered closed.

**Comment by point-source** [comment]
But that would be something being executed within the container, right? And shouldn't a brand new container be the same across machines? How is it possible that the script creates a different container on my system than it does on yours?

**Comment by tteck** [comment]
Was the command executed directly within the Proxmox shell?

If you are accessing a remote computer using SSH or a similar method, the shell session might be influenced by the local system's configuration, which would differ from your remote system.

**Comment by point-source** [comment]
I am connected via wireguard tunnel to the LAN that the proxmox server is physically attached to. I am visiting the proxmox management web ui by going to the IP address of the proxmox server with a Firefox browser session that has its cache and cookies cleared. I use the UI to select the proxmox node in the left side menu and then choose "Console". This opens a shell which is currently set to run fish but I have switched to bash in order to run your script.

Is this the right way to execute this or are you saying it should be physically with a k/v/m in front of the bare metal server?

EDIT: FWIW, I am currently setting up a proxmox instance in a vm to try to isolate the issue / reduce variables.

EDIT 2: Sure enough, worked in the VM. Very strange. The one exhibiting issues is a new proxmox deployment done two days ago. It has not been modified or customized other than the import of a zfs pool and the creation of a couple rocky linux containers and templates. I can't imagine that would affect it but I'll continue to search.

**Comment by tteck** [comment]
Establish a connection to Proxmox using the IP within the local network, bypassing the use of WireGuard.

**Comment by point-source** [comment]
> Establish a connection to Proxmox using the IP within the local network, bypassing the use of WireGuard.

That will have to wait since I am not at the physical site nor do I really intend to be very often. That said, I've worked around it for now by just installing nvm manually while the script executes and benefiting from the race condition.

In case you didn't catch my second edit to my last post, a new install of proxmox on a vm does work, confirming your previous output. So something about this deployment is not working but it's not really modified so I have no idea what it is. Thanks for the quick responses.

**Comment by tteck** [comment]
Given your existing connection method, you might encounter this issue every time the `.bashrc` file needs to be reloaded. Fortunately, only a handful of scripts necessitate this action.

**Comment by fredimachado** [comment]
Hey, I was having the exact same issue. I tried everything, even connecting a keyboard and a monitor to the server, it always failed in the nvm part.
The only way I was able to proceed was by manually installing nvm while the script is running (I had to try executing it multiple times, because curl and git are required and were being installed), before the Node.Js part. Thanks @point-source !
Thanks for your scripts @tteck !
